src/opustrainer/logger.py

```diff
@@ -0,0 +1,44 @@
+from io import TextIOWrapper
+import logging
+from sys import stderr
+from typing import List, TextIO
+from functools import lru_cache
+
+def getLogLevel(name: str) -> int:
+    """Incredibly, i can't find a function that will do this conversion, other
+    than setLevel, but setLevel doesn't work for calling the different log level logs."""
+    if name.upper() in logging.getLevelNamesMapping():
+        return logging.getLevelNamesMapping()[name.upper()]
+    else:
+        logging.log(logging.WARNING, "unknown log level level used: " + name + " assuming warning...")
+        return logging.WARNING
+
+def log(msg: str, loglevel: str = "INFO") -> None:
+    level = getLogLevel(loglevel)
+    logging.log(level, msg)
+
+
+@lru_cache(None)
+def log_once(msg: str, loglevel: str = "INFO") -> None:
+    """A wrapper to log, to make sure that we only print things once"""
+    log(msg, loglevel)
+
+
+def setup_logger(outputfilename: str | None = None, loglevel: str = "INFO", disable_stderr: bool=False) -> None:
+    """Sets up the logger with the necessary settings. Outputs to both file and stderr"""
+    loggingformat = '[%(asctime)s] [Trainer] [%(levelname)s] %(message)s'
+    handlers: List[logging.StreamHandler[TextIO] | logging.StreamHandler[TextIOWrapper]] = []
+    # disable_stderr is to be used only when testing the logger
```

Comment:
*grumpy noises*

Reply:
My type checker flagged it as an incompatible type assignment and this was the inferred type. I guess UNION should be used instead.

Reply:
For the WARNING I thought it'd be easier to use a text based log level from modifiers, in case users want to add their own logging, rather than getting to python internals.

That being said, `logging.getLevelNamesMapping()` is only available in python 3.11 so I shouldn't be using it.

Reply:
Also, since we define the level at runtime, I had to have `str` -> `loglevel` conversion anyways..